mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,266 @@
+/* Copyright (c) Fortanix, Inc.
+ *
+ * Licensed under the GNU General Public License, version 2 <LICENSE-GPL or
+ * https://www.gnu.org/licenses/gpl-2.0.html> or the Apache License, Version
+ * 2.0 <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0>, at your
+ * option. This file may not be copied, modified, or distributed except
+ * according to those terms. */
+
+#![cfg(all(feature = "std", feature = "async"))]
+
+use crate::{
+    error::{Error, Result},
+    ssl::{
+        context::Context,
+        io::IoCallback,
+    },
+};
+use async_trait::async_trait;
+use std::{
+    cell::Cell,
+    future::Future,
+    io::{Error as IoError, ErrorKind as IoErrorKind, Result as IoResult},
+    pin::Pin,
+    ptr::null_mut,
+    rc::Rc,
+    result::Result as StdResult,
+    task::{Context as TaskContext, Poll},
+};
+use tokio::{
+    io::{AsyncRead, AsyncWrite, ReadBuf},
+    net::UdpSocket,
+};
+
+#[derive(Clone)]
+pub struct ErasedContext(Rc<Cell<*mut ()>>);
```

Comment:
What does this do?

Reply:
Added comments

Reply:
This type is a wrapper for arbitrary [`std::task::Context`]. It's used with
[`AsyncIoAdapter`] to pass the context of current async task into [`IoCallback`]
so that async context could be access in the callback inside the `mbedtls`